### PR TITLE
Cleanup compression

### LINKED ISSUE #12688: CLN: consolidate compression inference as much as possible
we have 3 copies of the compression inference routines in `parser.py, parser.py, and io/common.py`. See how much we can reduce this (_some_ is necessary as sometimes we need to actually return a file handle or need to read in and re-wrap).

xref #12175

### PR DESCRIPTION
**Pull request by lababidi**
- [x] closes #12688 
- [x] tests passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry - not needed

### CODE REVIEW

File: pandas/io/common.py [hunk 1]
```diff
@@ -276,56 +276,92 @@ def ZipFile(*args, **kwargs):
...
 
-def _get_handle(path, mode, encoding=None, compression=None):
-    """Gets file handle for given path and mode.
+def _get_handle(source, mode, encoding=None, compression=None):
     """
-    if compression is not None:
-        if encoding is not None and not compat.PY3:
+    Gets file handle for given source and mode.
+    wraps compressed fileobject in a decompressing fileobject
+    NOTE: For all files in Python 3.2 and for bzip'd files under all Python
```
Comment (jreback) [inline pandas/io/common.py]: we want to remove this statement and its functionaility as we don't support 3.2 anymore

File: pandas/io/parsers.py [hunk 2]
```diff
@@ -1652,16 +1651,7 @@ def __init__(self, f, **kwds):
         self.comment = kwds['comment']
         self._comment_lines = []
 
-        if isinstance(f, compat.string_types):
-            f = _get_handle(f, 'r', encoding=self.encoding,
-                            compression=self.compression)
-        elif self.compression:
```
Comment (jreback) [inline pandas/io/parsers.py]: you can remove the ``_wrap_compressed`` function

**Comment by lababidi** [comment]
@jreback how is it possible that i broke the following? I don't think I touched any of these files:

```Traceback (most recent call last):
  File "/Users/travis/build/pydata/pandas/pandas/tests/frame/test_to_csv.py", line 999, in test_to_csv_compression_xz
    df.to_csv(filename, compression="xz")
  File "/Users/travis/build/pydata/pandas/pandas/core/frame.py", line 1344, in to_csv
    formatter.save()
  File "/Users/travis/build/pydata/pandas/pandas/formats/format.py", line 1551, in save
    self._save()
  File "/Users/travis/build/pydata/pandas/pandas/formats/format.py", line 1638, in _save
    self._save_header()
  File "/Users/travis/build/pydata/pandas/pandas/formats/format.py", line 1634, in _save_header
    writer.writerow(encoded_labels)
io.UnsupportedOperation: not writable```

**Comment by jreback** [comment]
well that writes to a csv so sure

**Comment by lababidi** [comment]
But it never calls get_handle??

All the tests pass on my machine. This is strange.

**Comment by jreback** [comment]
are you sure that test is running? it will skip if you don't have ``backports.lzma`` not installed

**Comment by lababidi** [comment]
yeah, i installed `backports.lzma` and the test runs on my machine. this is strange. especially since i didn't touch any of the files in the stacktrace.

**Comment by jreback** [comment]
try creating the exact environment (e.g. create a new environment with ``ci/requirements-2.7.run/pip``)

**Comment by lababidi** [comment]
how do i install the `.run` file? `pip` needs `==` while the `.run` file has `=`

**Comment by jreback** [comment]
```
conda install -n pandas --file=ci/requirements.....run
pip install --upgrade -r ci/requirements....pip
```

**Comment by jreback** [comment]
can you rebase/update

**Comment by jreback** [comment]
can you update

**Comment by lababidi** [comment]
This branch is behind on master. `memory_map` was added and I need time to understand how that changes the logic.

**Comment by jorisvandenbossche** [comment]
I think something went wrong with your rebase (lots of commits are included here)

**Comment by jorisvandenbossche** [comment]
@lababidi Do you have time to update this? (you should start with a rebase, as lots of other commits are not included in this PR)

**Comment by jreback** [comment]
can you rebase / update?

**Comment by lababidi** [comment]
@jreback on it now

**Comment by lababidi** [comment]
@jreback @jorisvandenbossche Again, after rebasing to your code, I find that the tests fail and not due to my changes. This is becoming a detriment to my contributions because it seems someone keeps allowing failing code to get merged. I'm not sure how it happened.

**Comment by lababidi** [comment]
@jreback @jorisvandenbossche I've given you access to modify this branch, so feel free to modify anything you'd like. Thank you.

**Comment by jreback** [comment]
are these your changes: https://travis-ci.org/pydata/pandas/jobs/159174397

master has been and is passing continuously. any failures are changes that you made. They can certainly fail other tests, but that just means your changes are not clean.

**Comment by jorisvandenbossche** [comment]
@lababidi If you look at the test failures in the log @jreback linked to, you see that the failure is in the `test_iterators` test (in `test_parsers.py`, so related to this PR). 
The test failes at `chunks = list(parser)`, and this `parser` is a file reader with a `PythonParser` engine which is passed a list of list of strings instead of a file-like object. Because of your change, this list is added to the `engine.handles`, and this gives the error you see in the test failure "'list' object has no attribute 'close'"

**Comment by lababidi** [comment]
@jreback @jorisvandenbossche Thank you for the insight. The `handles` is new to me. I must be passing the wrong thing through. I'll have to understand the logic correctly on what we're aiming for.

**Comment by dhimmel** [comment]
What is the remaining work that needs to be done for this pull request? I'm willing to give it a try, but need a little guidance to get started.

**Comment by jorisvandenbossche** [comment]
@dhimmel To start with, you can look into the failing tests. As that is of course the minimum for getting this merged. 
The failure has to do with passing a list of lines to `TextParser`, which now no longer works.

**Pull request closed without merging**